File: Java-Versions/Question2/src/Question2.java

/*
    Question2
    Java
    Adam Giaccaglia

 */

public class Question2 {
    public static void main( String[] args ) {
        Musician guy = new Musician("billy");

        guy.sing();
        guy.pianoPlay();
        guy.violinPlay();
    }
}

class Pianist
{
    private String name;
    public Pianist(String x)
    {
        name = x;
    }
    public void pianoPlay(){
        System.out.println("Play a piano!");
    }
}

class Violinist implements IViolinist{
    public void violinPlay() {
            System.out.println("Play a violin!");
    }

}

interface IViolinist
{
    void violinPlay();
}


class Musician extends Pianist implements IViolinist{

    public Musician(String nm) {
             super(nm);
    }
    public void sing() {
            System.out.println("Sing a song!");
    }

    // makes having a violinist class redundant
    public void violinPlay() {
            System.out.println("Play a violin!");
    }
}

File: C PLusPlus Versions/Question2/Question2.cpp
// Question2.cpp : Defines the entry point for the console application.
//

/*
	Question 2
	C++
	Adam Giaccaglia
*/

#include "stdafx.h"
#include <iostream>
#include <string>

using namespace std;

class Pianist
{
//protected to allow pianist to inherit
protected:
	string name;
public:
	Pianist(string x)
	{
		name = x;
	}
 void pianoPlay(){
	 cout << ("Play a piano!") << endl;
	}
};

class Violinist
{
public:
	void violinPlay()
	{
		cout << ("Play a violin!") << endl;
	}
};

class Musician: public Pianist, public Violinist
{
public:
	Musician(string str) :Pianist(str), Violinist(){

	}
	void sing()
	{
		cout << ("Sing a song!") << endl;
	}
};

int _tmain(int argc, _TCHAR* argv[])
{
	
	Musician * guy = new Musician("Bob");
	guy->sing();
	guy->pianoPlay();
	guy->violinPlay();
	
	return 0;
}

